Add a mechanism for on-demand logging about where a process is loading its code and resources from

davidquesada opened the pull request:
#### abe18ff38594845ea1d3cdf4dc8a747f0ddccd93
<pre>
Add a mechanism for on-demand logging about where a process is loading its code and resources from
<a href="https://bugs.webkit.org/show_bug.cgi?id=243458">https://bugs.webkit.org/show_bug.cgi?id=243458</a>
rdar://91325510

Reviewed by Saam Barati.

In order to verify with certainty that WebKit clients are using all the correct content (i.e.
loading code from the correct framework bundles or DYLD shared cache, and loading resources
from the correct paths), it can be useful for a process to self-report all these details so
that automated tests (or manual visual inspection) can validate the coherency of the WebKit
stack (as well as any other application frameworks that may need to be loaded from a variety
of potential locations).

To support this, introduce a means by which clients of the WebKit stack (or more precisely,
users of WTF via JavaScriptCore) can automatically log this information on demand. Upon
initialization of WTF, processes will observe a distributed notification. Upon receiving this
notification, the process will introspect the DYLD shared cache, a specified set of dynamic
libraries, and a specified set of bundles. Information about those items will then be printed
to a dedicated OS log channel as a series of JSON outout that external tooling can parse for
consumption by humans or further tooling.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
    Add references to LibraryPathDiagnostics.{h,mm} and CFPrivSPI.h.
* Source/WTF/wtf/PlatformHave.h:
    Add a few conditional flags for SPI used to gather info to be logged.
* Source/WTF/wtf/Threading.cpp:
(WTF::initialize):
    Initialize the new &quot;library path diagnostics&quot; system when setting up WTF.
* Source/WTF/wtf/darwin/LibraryPathDiagnostics.h: Added.
* Source/WTF/wtf/darwin/LibraryPathDiagnostics.mm: Added.
(WTF::LibraryPathDiagnosticsLogger::LibraryPathDiagnosticsLogger):
(WTF::LibraryPathDiagnosticsLogger::logJSONPayload):
    Logs an individual piece of information by writing the stringification of a JSON payload
    to the associated OS log channel. Examples of payloads could be the UUID/path of the
    process&apos;s DYLD shared cache, or the location and version of one loaded bundle with a
    given bundle identifier. The overall compilation of information produced by this logger
    is divided and delivered as multiple JSON payloads because the OS log places a limit on
    the size of a message at 1024 bytes -- a limit that could easily be exceeded if a process
    were to just log the UUIDs and full dylib / bundle paths of every library in the WebKit
    stack. Tooling that wishes to analyze this information must merge all the individual JSON
    payloads read from the system log in order to assemble the complete understanding of a
    process&apos;s dynamic library stack.
(WTF::LibraryPathDiagnosticsLogger::logString):
    Emits a string to the log, at a given &quot;path&quot; of nested objects, e.g. logging &quot;XYZ&quot; at the
    path &quot;Foo&quot;, &quot;Bar&quot;, &quot;Baz&quot; would print the JSON payload `{&quot;Foo&quot;:{&quot;Bar&quot;:{&quot;Baz&quot;:&quot;XYZ&quot;}}}`.
(WTF::LibraryPathDiagnosticsLogger::logObject):
    Ditto, but prints an object at a specified path.
(WTF::LibraryPathDiagnosticsLogger::logError):
    Writes the specified string to the OS log as an error.
(WTF::LibraryPathDiagnosticsLogger::logExecutablePath):
    Logs the current executable&apos;s path, after canonicalizing it by resolving all symlinks.
(WTF::LibraryPathDiagnosticsLogger::logDYLDSharedCacheInfo):
    Logs the UUID and path on disk of the process&apos;s active DYLD shared cache.
(WTF::isAddressInSharedRegion):
(WTF::LibraryPathDiagnosticsLogger::logDynamicLibraryInfo):
    Logs information about a particular dynamic library, specified by the install name. Using
    `dlopen()` to grab a handle to the library (only if it&apos;s already loaded), read the UUID,
    the path &quot;on disk&quot; to the library, and determine whether or not the library (referred to
    by its mach header) is present in the shared cache.
(WTF::LibraryPathDiagnosticsLogger::logBundleInfo):
    Logs information about a particular (Core)Foundation bundle, specified by bundle identifier.
    Look up the bundle and log its version and path on disk (after resolving all symlinks).
(WTF::LibraryPathDiagnosticsLogger::logCryptexCanaryInfo):
    Use chirp to look up canary info for the OS and App cryptexes.
(WTF::LibraryPathDiagnosticsLogger::log):
    Invoke each of the respective methods to log the various categories of information.
(WTF::initializeLibraryPathDiagnostics):
    Register for the distributed notification that triggers the process to log its library info.
(WTF::logLibraryPathDiagnostics):
* Source/WTF/wtf/spi/cf/CFPrivSPI.h: Added.
* Source/WTF/wtf/spi/darwin/dyldSPI.h:

Canonical link: <a href="https://commits.webkit.org/253455@main">https://commits.webkit.org/253455@main</a>
</pre>

saambarati commented on Source/WTF/wtf/darwin/LibraryPathDiagnostics.mm:
```diff
@@ -0,0 +1,301 @@
+    auto payload = JSON::Object::create();
+    auto iter = path.rbegin();
+    payload->setString(*iter, string);
+    for (iter++; iter != path.rend(); iter++) {
```
Can we use modern `for (:)` iteration here?

saambarati replied:
Oh I see, you're using the iterator once before the loop. This is a nit, but I probably would've used an indexed loop instead of this.

saambarati commented on Source/WTF/wtf/darwin/LibraryPathDiagnostics.mm:
```diff
@@ -0,0 +1,301 @@
+    auto payload = JSON::Object::create();
+    auto iter = path.rbegin();
+    payload->setObject(*iter, object);
+    for (iter++; iter != path.rend(); iter++) {
```
Can we use modern `for (:)` iteration here?

saambarati commented on Source/WTF/wtf/darwin/LibraryPathDiagnostics.mm:
```diff
@@ -0,0 +1,301 @@
+{
+    auto payload = JSON::Object::create();
+    auto iter = path.rbegin();
+    payload->setObject(*iter, object);
```
I think you can use `WTFMove(object)` here

saambarati replied:
Which also makes it less weird to take an rvalue ref but not move it.

saambarati reviewed (approved):
r=me

saambarati commented on Source/WTF/wtf/darwin/LibraryPathDiagnostics.mm:
```diff
@@ -0,0 +1,301 @@
+#if HAVE(SHARED_REGION_SPI)
+static bool isAddressInSharedRegion(const void* addr)
+{
+    return (uintptr_t)addr >= SHARED_REGION_BASE && (uintptr_t)addr < (SHARED_REGION_BASE + SHARED_REGION_SIZE);
```
nit: let's use static_cast or bitwise_cast

davidquesada committed:
Add a mechanism for on-demand logging about where a process is loading its code and resources from
https://bugs.webkit.org/show_bug.cgi?id=243458
rdar://91325510

Reviewed by Saam Barati.

In order to verify with certainty that WebKit clients are using all the correct content (i.e.
loading code from the correct framework bundles or DYLD shared cache, and loading resources
from the correct paths), it can be useful for a process to self-report all these details so
that automated tests (or manual visual inspection) can validate the coherency of the WebKit
stack (as well as any other application frameworks that may need to be loaded from a variety
of potential locations).

To support this, introduce a means by which clients of the WebKit stack (or more precisely,
users of WTF via JavaScriptCore) can automatically log this information on demand. Upon
initialization of WTF, processes will observe a distributed notification. Upon receiving this
notification, the process will introspect the DYLD shared cache, a specified set of dynamic
libraries, and a specified set of bundles. Information about those items will then be printed
to a dedicated OS log channel as a series of JSON outout that external tooling can parse for
consumption by humans or further tooling.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
    Add references to LibraryPathDiagnostics.{h,mm} and CFPrivSPI.h.
* Source/WTF/wtf/PlatformHave.h:
    Add a few conditional flags for SPI used to gather info to be logged.
* Source/WTF/wtf/Threading.cpp:
(WTF::initialize):
    Initialize the new "library path diagnostics" system when setting up WTF.
* Source/WTF/wtf/darwin/LibraryPathDiagnostics.h: Added.
* Source/WTF/wtf/darwin/LibraryPathDiagnostics.mm: Added.
(WTF::LibraryPathDiagnosticsLogger::LibraryPathDiagnosticsLogger):
(WTF::LibraryPathDiagnosticsLogger::logJSONPayload):
    Logs an individual piece of information by writing the stringification of a JSON payload
    to the associated OS log channel. Examples of payloads could be the UUID/path of the
    process's DYLD shared cache, or the location and version of one loaded bundle with a
    given bundle identifier. The overall compilation of information produced by this logger
    is divided and delivered as multiple JSON payloads because the OS log places a limit on
    the size of a message at 1024 bytes -- a limit that could easily be exceeded if a process
    were to just log the UUIDs and full dylib / bundle paths of every library in the WebKit
    stack. Tooling that wishes to analyze this information must merge all the individual JSON
    payloads read from the system log in order to assemble the complete understanding of a
    process's dynamic library stack.
(WTF::LibraryPathDiagnosticsLogger::logString):
    Emits a string to the log, at a given "path" of nested objects, e.g. logging "XYZ" at the
    path "Foo", "Bar", "Baz" would print the JSON payload `{"Foo":{"Bar":{"Baz":"XYZ"}}}`.
(WTF::LibraryPathDiagnosticsLogger::logObject):
    Ditto, but prints an object at a specified path.
(WTF::LibraryPathDiagnosticsLogger::logError):
    Writes the specified string to the OS log as an error.
(WTF::LibraryPathDiagnosticsLogger::logExecutablePath):
    Logs the current executable's path, after canonicalizing it by resolving all symlinks.
(WTF::LibraryPathDiagnosticsLogger::logDYLDSharedCacheInfo):
    Logs the UUID and path on disk of the process's active DYLD shared cache.
(WTF::isAddressInSharedRegion):
(WTF::LibraryPathDiagnosticsLogger::logDynamicLibraryInfo):
    Logs information about a particular dynamic library, specified by the install name. Using
    `dlopen()` to grab a handle to the library (only if it's already loaded), read the UUID,
    the path "on disk" to the library, and determine whether or not the library (referred to
    by its mach header) is present in the shared cache.
(WTF::LibraryPathDiagnosticsLogger::logBundleInfo):
    Logs information about a particular (Core)Foundation bundle, specified by bundle identifier.
    Look up the bundle and log its version and path on disk (after resolving all symlinks).
(WTF::LibraryPathDiagnosticsLogger::logCryptexCanaryInfo):
    Use chirp to look up canary info for the OS and App cryptexes.
(WTF::LibraryPathDiagnosticsLogger::log):
    Invoke each of the respective methods to log the various categories of information.
(WTF::initializeLibraryPathDiagnostics):
    Register for the distributed notification that triggers the process to log its library info.
(WTF::logLibraryPathDiagnostics):
* Source/WTF/wtf/spi/cf/CFPrivSPI.h: Added.
* Source/WTF/wtf/spi/darwin/dyldSPI.h:

Canonical link: https://commits.webkit.org/253455@main
